README.md
# `depensort`

`a package that helps you keep a beautiful looking package.json that has been sorted as per your configuration`

[![Documentation site](https://img.shields.io/static/v1?style=for-the-badge&label=%20&message=documentation&color=blue&logo=readthedocs&logoColor=white)](https://npm.officialk.dev/depensort)

**NPM**

[![NPM version](https://img.shields.io/npm/v/depensort?style=for-the-badge)](https://npmjs.org/package/depensort)
![NPM size](https://img.shields.io/bundlephobia/min/depensort?style=for-the-badge)
![NPM vunerabilities](https://img.shields.io/snyk/vulnerabilities/npm/depensort?style=for-the-badge)

**GitHub**

[![GitHub License](https://img.shields.io/github/license/npm-officialk/depensort?style=for-the-badge)](https://github.com/npm-officialk/depensort/README.md)
[![GitHub issues](https://img.shields.io/github/issues/npm-officialk/depensort?style=for-the-badge)](https://github.com/npm-officialk/depensort/issues/)
[![GitHub PRs](https://img.shields.io/github/issues-pr/npm-officialk/depensort?style=for-the-badge)](https://github.com/npm-officialk/depensort/pulls/)
[![GitHub Milestones](https://img.shields.io/github/milestones/all/npm-officialk/depensort?style=for-the-badge)](https://github.com/npm-officialk/depensort/milestones/)
![GitHub Sponsors](https://img.shields.io/github/sponsors/npm-officialk?style=for-the-badge)

## Installation

```yarn
yarn add depensort --dev
```

```npm
npm i -D depensort
```

## Usage

The package is used to sort the dependencies in the package.json file in ascending order of key length after every install(or on command)

### 1. Set the package to run after every install

```json
"scripts":{
	"postinstall":"depensort"
}
```

### 2. Call it manually

**From the cmd**

-   Install the package globally

```npm
npm i -g depensort
```

OR

```yarn
yarn global add depensort
```

-   Call it from the cmd

```bash
depensort
```

## For the devs

[Refer to the Developer Release Documentation](https://npm-officialk.github.io/depensort)

use the `yarn link` in the package's folder

and `yarn link "depensort"` in the folder you want to test the package in

this will install the package as a symbolic link in the test folder

now you can now use the package as if it was installed.

## Contributing

Pull requests are welcome. For major changes, please open an issue first
to discuss what you would like to change.

Some files to refer

-   [CONTRIBUTING](./.github/CONTRIBUTING.md)
-   [CODE_OF_CONDUCT](./.github/CODE_OF_CONDUCT.md)

Please make sure to update tests as appropriate.

## License

[LICENSE](./LICENSE)


CHANGELOG.md
## 1.0.0 (2023-07-09)

* refactor(package): updated package to use the yargs package ([7467714](https://github.com/npm-officialk/depensort/commit/7467714))
* feat(v1): releasing v1 ([9d1e0f8](https://github.com/npm-officialk/depensort/commit/9d1e0f8)), closes [#8](https://github.com/npm-officialk/depensort/issues/8) [#8](https://github.com/npm-officialk/depensort/issues/8) [#11](https://github.com/npm-officialk/depensort/issues/11) [#9](https://github.com/npm-officialk/depensort/issues/9) [#11](https://github.com/npm-officialk/depensort/issues/11) [#9](https://github.com/npm-officialk/depensort/issues/9) [#10](https://github.com/npm-officialk/depensort/issues/10) [#12](https://github.com/npm-officialk/depensort/issues/12) [#10](https://github.com/npm-officialk/depensort/issues/10) [#13](https://github.com/npm-officialk/depensort/issues/13) [#15](https://github.com/npm-officialk/depensort/issues/15) [#16](https://github.com/npm-officialk/depensort/issues/16)
* chore(setup): Intial commit [skip ci] ([52dda32](https://github.com/npm-officialk/depensort/commit/52dda32))


### BREAKING CHANGE

* removed the export of the function and instead calling the function

## 1.0.0-development.3 (2023-07-08)

-   refactor(package): updated depensort to use yargs package ([c18cb9c](https://github.com/npm-officialk/depensort/commit/c18cb9c))
-   19 feature upgrade to yargs (#20) ([5901313](https://github.com/npm-officialk/depensort/commit/5901313)), closes [#20](https://github.com/npm-officialk/depensort/issues/20) [#19](https://github.com/npm-officialk/depensort/issues/19)
-   ci: updated prettier action version ([904f49a](https://github.com/npm-officialk/depensort/commit/904f49a))
-   ci: updated schedules, release and husky ([a2a5b67](https://github.com/npm-officialk/depensort/commit/a2a5b67)), closes [#16](https://github.com/npm-officialk/depensort/issues/16)
-   ci(formatting): [skip ci] updating format to be consistent ([514173c](https://github.com/npm-officialk/depensort/commit/514173c))
-   test(package): updated the integration/package level tests to be more… (#13) ([115bd10](https://github.com/npm-officialk/depensort/commit/115bd10)), closes [#13](https://github.com/npm-officialk/depensort/issues/13) [#15](https://github.com/npm-officialk/depensort/issues/15)
-   docs(docs): resolved bug in documentation #10 (#12) ([e8042f3](https://github.com/npm-officialk/depensort/commit/e8042f3)), closes [#10](https://github.com/npm-officialk/depensort/issues/10) [#12](https://github.com/npm-officialk/depensort/issues/12) [#10](https://github.com/npm-officialk/depensort/issues/10)

### BREAKING CHANGE

-   updated the entry point to use yargs

## 1.0.0-development.3 (2023-07-08)

-   ci: updated prettier action version ([904f49a](https://github.com/npm-officialk/depensort/commit/904f49a))
-   ci: updated quation ([6309718](https://github.com/npm-officialk/depensort/commit/6309718))
-   ci: updated schedules, release and husky ([a2a5b67](https://github.com/npm-officialk/depensort/commit/a2a5b67)), closes [#16](https://github.com/npm-officialk/depensort/issues/16)
-   ci(formatting): [skip ci] updating format to be consistent ([514173c](https://github.com/npm-officialk/depensort/commit/514173c))
-   docs: updated a spelling error ([c67c071](https://github.com/npm-officialk/depensort/commit/c67c071))
-   docs(docs): resolved bug in documentation #10 (#12) ([e8042f3](https://github.com/npm-officialk/depensort/commit/e8042f3)), closes [#10](https://github.com/npm-officialk/depensort/issues/10) [#12](https://github.com/npm-officialk/depensort/issues/12) [#10](https://github.com/npm-officialk/depensort/issues/10)
-   19 feature upgrade to yargs (#20) ([5901313](https://github.com/npm-officialk/depensort/commit/5901313)), closes [#20](https://github.com/npm-officialk/depensort/issues/20) [#19](https://github.com/npm-officialk/depensort/issues/19)
-   test(package): updated the integration/package level tests to be more… (#13) ([115bd10](https://github.com/npm-officialk/depensort/commit/115bd10)), closes [#13](https://github.com/npm-officialk/depensort/issues/13) [#15](https://github.com/npm-officialk/depensort/issues/15)

### BREAKING CHANGE

-   updated the entry point to use yargs

## 1.0.0-development.2 (2023-02-17)

-   fix(package): updated response to call function instead of exporting it (#11) ([1acfc3a](https://github.com/npm-officialk/depensort/commit/1acfc3a)), closes [#11](https://github.com/npm-officialk/depensort/issues/11) [#9](https://github.com/npm-officialk/depensort/issues/9)
-   ci(formatting): [skip ci] updating format to be consistent ([768b7b6](https://github.com/npm-officialk/depensort/commit/768b7b6))

### BREAKING CHANGE

-   removed the export of the function and instead calling the function

## 1.0.0-development.1 (2023-02-15)

-   fix(cicd): added env variables to the release action ([39cd516](https://github.com/npm-officialk/depensort/commit/39cd516))
-   feat(mvp): added sorting feature (#8) ([9c43f80](https://github.com/npm-officialk/depensort/commit/9c43f80)), closes [#8](https://github.com/npm-officialk/depensort/issues/8)
-   chore(setup): Intial commit [skip ci] ([52dda32](https://github.com/npm-officialk/depensort/commit/52dda32))


tests/index.spec.js
const { execSync } = require("child_process")
const { readFileSync, writeFileSync } = require("fs")
let contents
describe("Testing the working of the package", () => {
	beforeAll(() => {
		contents = JSON.parse(readFileSync("package.json", "utf8"))
	})
	afterAll(() => {
		writeFileSync("package.json", JSON.stringify(contents, null, 4), "utf8")
	})
	it("should sort package json devDependencies", () => {
		execSync("yarn run sort")
		const devDependencies = contents.devDependencies
		const sortedArray = Object.entries(devDependencies).sort((first, second) => {
			let difference = first[0].length - second[0].length
			if (difference === 0) {
				difference = first[1].length - second[1].length
			}
			return difference
		})
		const newFile = JSON.parse(readFileSync("package.json", "utf8"))
		expect(JSON.stringify(Object.fromEntries(sortedArray))).toBe(JSON.stringify(newFile.devDependencies))
	})
})


tests/jest.config.js
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/configuration
 */

module.exports = {
	displayName: { name: "depensort-integration testing", color: "green" },
	// All imported modules in your tests should be mocked automatically
	// automock: false,

	// Stop running tests after `n` failures
	// bail: 0,

	// The directory where Jest should store its cached dependency information
	// cacheDirectory: "C:\\Users\\karkoppa\\AppData\\Local\\Temp\\jest",

	// Automatically clear mock calls, instances, contexts and results before every test
	clearMocks: true,

	// Indicates whether the coverage information should be collected while executing the test
	// collectCoverage: true,

	// An array of glob patterns indicating a set of files for which coverage information should be collected
	// collectCoverageFrom: undefined,

	// The directory where Jest should output its coverage files
	// coverageDirectory: "./coverage",

	// An array of regexp pattern strings used to skip coverage collection
	// coveragePathIgnorePatterns: ["\\\\node_modules\\\\"],

	// Indicates which provider should be used to instrument code for coverage
	// coverageProvider: "v8",

	// A list of reporter names that Jest uses when writing coverage reports
	// coverageReporters: ["json", "lcov"],

	// An object that configures minimum threshold enforcement for coverage results
	// coverageThreshold: {
	// 	global: {
	// 		branches: 80,
	// 		functions: 80,
	// 		lines: 80,
	// 		statements: -10,
	// 	},
	// },

	// A path to a custom dependency extractor
	// dependencyExtractor: undefined,

	// Make calling deprecated APIs throw helpful error messages
	// errorOnDeprecated: false,

	// The default configuration for fake timers
	// fakeTimers: {
	//   "enableGlobally": false
	// },

	// Force coverage collection from ignored files using an array of glob patterns
	// forceCoverageMatch: [],

	// A path to a module which exports an async function that is triggered once before all test suites
	// globalSetup: undefined,

	// A path to a module which exports an async function that is triggered once after all test suites
	// globalTeardown: undefined,

	// A set of global variables that need to be available in all test environments
	// globals: {},

	// The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
	// maxWorkers: "50%",

	// An array of directory names to be searched recursively up from the requiring module's location
	// moduleDirectories: [
	//   "node_modules"
	// ],

	// An array of file extensions your modules use
	moduleFileExtensions: ["js", "ts", "json"],

	// A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
	// moduleNameMapper: {},

	// An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
	// modulePathIgnorePatterns: [],

	// Activates notifications for test results
	// notify: false,

	// An enum that specifies notification mode. Requires { notify: true }
	// notifyMode: "failure-change",

	// A preset that is used as a base for Jest's configuration
	// preset: undefined,

	// Run tests from one or more projects
	// projects: undefined,

	// Use this configuration option to add custom reporters to Jest
	// reporters: undefined,

	// Automatically reset mock state before every test
	// resetMocks: false,

	// Reset the module registry before running each individual test
	// resetModules: false,

	// A path to a custom resolver
	// resolver: undefined,

	// Automatically restore mock state and implementation before every test
	// restoreMocks: false,

	// The root directory that Jest should scan for tests and modules within
	// rootDir: "tests",

	// A list of paths to directories that Jest should use to search for files in
	roots: ["."],

	// Allows you to use a custom runner instead of Jest's default test runner
	// runner: "jest-runner",

	// The paths to modules that run some code to configure or set up the testing environment before each test
	// setupFiles: [],

	// A list of paths to modules that run some code to configure or set up the testing framework before each test
	// setupFilesAfterEnv: [],

	// The number of seconds after which a test is considered as slow and reported as such in the results.
	// slowTestThreshold: 5,

	// A list of paths to snapshot serializer modules Jest should use for snapshot testing
	// snapshotSerializers: [],

	// The test environment that will be used for testing
	// testEnvironment: "jest-environment-node",

	// Options that will be passed to the testEnvironment
	// testEnvironmentOptions: {},

	// Adds a location field to test results
	// testLocationInResults: false,

	// The glob patterns Jest uses to detect test files
	testMatch: ["**/?(*.)+(spec|test).[tj]s?(x)"],

	// An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
	testPathIgnorePatterns: ["\\\\node_modules\\\\"],

	// The regexp pattern or array of patterns that Jest uses to detect test files
	// testRegex: [],

	// This option allows the use of a custom results processor
	// testResultsProcessor: undefined,

	// This option allows use of a custom test runner
	// testRunner: "jest-circus/runner",

	// A map from regular expressions to paths to transformers
	transform: {
		"^.+\\.(ts|tsx)$": "ts-jest",
	},

	// An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
	// transformIgnorePatterns: [
	//   "\\\\node_modules\\\\",
	//   "\\.pnp\\.[^\\\\]+$"
	// ],

	// An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
	// unmockedModulePathPatterns: undefined,

	// Indicates whether each individual test should be reported during the run
	// verbose: undefined,

	// An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
	// watchPathIgnorePatterns: [],

	// Whether to use watchman for file crawling
	// watchman: true,
}


src/File/File.spec.ts
import { existsSync, mkdirSync, writeFileSync, rmSync } from "fs"
import File from "./"

const testAssetsDir = `./.cache`

const goodJson = {
	file: "used for testing purposes and should be deleted automatically",
	usedFor: "testing",
	deleted: false,
}

const notJson = "used for testing purposes and should be deleted automatically"

const setContent = () => {
	if (!existsSync(testAssetsDir)) {
		mkdirSync(testAssetsDir)
	}
	writeFileSync(`${testAssetsDir}/test.json`, JSON.stringify(goodJson), "utf8")
	writeFileSync(`${testAssetsDir}/.testrc`, JSON.stringify(goodJson), "utf8")
	writeFileSync(`${testAssetsDir}/test-fail.json`, notJson, "utf8")
	writeFileSync(`${testAssetsDir}/.test-failrc`, notJson, "utf8")
}

const clearFiles = () => {
	rmSync(`${testAssetsDir}/test.json`)
	rmSync(`${testAssetsDir}/.testrc`)
	rmSync(`${testAssetsDir}/test-fail.json`)
	rmSync(`${testAssetsDir}/.test-failrc`)
}

describe("Testing the File class and its methods", () => {
	beforeEach(() => {
		setContent()
	})
	afterAll(() => {
		clearFiles()
	})

	describe("Testing the constructor", () => {
		it("should create an instance when a valid file path is passed", () => {
			expect(new File(`${testAssetsDir}/test.json`)._file).toBe(`${testAssetsDir}/test.json`)
		})
		it("should throw an error when passed a invalid file path", () => {
			expect(() => {
				new File("./doesnt-exist.json")
			}).toThrow("File does not exist!")
		})
		it("should throw an error when passed a directory path", () => {
			expect(() => new File("./")).toThrow("Entered a directory, expected a file!")
		})
	})

	describe("Testing the read method", () => {
		it("should return json(parsed) content when reading a json file with valid content", () => {
			const file = new File(`${testAssetsDir}/test.json`)
			expect(file.read()).toStrictEqual(goodJson)
		})
		it("should return json(parsed) content when reading a non json file with valid content", () => {
			const file = new File(`${testAssetsDir}/.testrc`)
			expect(file.read()).toStrictEqual(goodJson)
		})
		it("should throw an error when reading a json file with bad/non json content", () => {
			const file = new File(`${testAssetsDir}/test-fail.json`)
			expect(() => {
				file.read()
			}).toThrowError()
		})
		it("should throw an error when reading a non json file with bad/non json content", () => {
			const file = new File(`${testAssetsDir}/.test-failrc`)
			expect(() => {
				file.read()
			}).toThrowError()
		})
	})

	describe("Testing the write method", () => {
		it("should write the json string to the file", () => {
			const file = new File(`${testAssetsDir}/.testrc`)
			file.write(JSON.stringify(goodJson))
			expect(file.read()).toStrictEqual(goodJson)
		})
		it("should write the json object to the file", () => {
			const file = new File(`${testAssetsDir}/.testrc`)
			file.write(goodJson)
			expect(file.read()).toStrictEqual(goodJson)
		})
	})
})


src/commands/all.ts
import File from "../utils/File"
import Sorter from "../utils/Sorter"

export const command = "$0"
export const alias = ["$0", "all"]
export const desc = "Sort all the dependencies in the package json"
export const builder = {}
export const handler = () => {
	const packageFile = new File(`${process.cwd()}/package.json`)
	const packageValues = packageFile.read()
	if (packageValues.dependencies) {
		packageValues.dependencies = new Sorter(packageValues.dependencies).sort()
	}
	if (packageValues.devDependencies) {
		packageValues.devDependencies = new Sorter(packageValues.devDependencies).sort()
	}
	if (packageValues.resolutions) {
		packageValues.resolutions = new Sorter(packageValues.resolutions).sort()
	}
	if (packageValues.overrides) {
		packageValues.overrides = new Sorter(packageValues.overrides).sort()
	}
	packageFile.write(packageValues)
}


jest.config.js
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/configuration
 */

module.exports = {
	displayName: { name: "depensort", color: "green" },
	// All imported modules in your tests should be mocked automatically
	// automock: false,

	// Stop running tests after `n` failures
	// bail: 0,

	// The directory where Jest should store its cached dependency information
	// cacheDirectory: "C:\\Users\\karkoppa\\AppData\\Local\\Temp\\jest",

	// Automatically clear mock calls, instances, contexts and results before every test
	clearMocks: true,

	// Indicates whether the coverage information should be collected while executing the test
	collectCoverage: true,

	// An array of glob patterns indicating a set of files for which coverage information should be collected
	// collectCoverageFrom: undefined,

	// The directory where Jest should output its coverage files
	coverageDirectory: "./coverage",

	// An array of regexp pattern strings used to skip coverage collection
	coveragePathIgnorePatterns: ["\\\\node_modules\\\\"],

	// Indicates which provider should be used to instrument code for coverage
	coverageProvider: "v8",

	// A list of reporter names that Jest uses when writing coverage reports
	coverageReporters: ["json", "lcov"],

	// An object that configures minimum threshold enforcement for coverage results
	coverageThreshold: {
		global: {
			branches: 80,
			functions: 80,
			lines: 80,
			statements: -10,
		},
	},

	// A path to a custom dependency extractor
	// dependencyExtractor: undefined,

	// Make calling deprecated APIs throw helpful error messages
	// errorOnDeprecated: false,

	// The default configuration for fake timers
	// fakeTimers: {
	//   "enableGlobally": false
	// },

	// Force coverage collection from ignored files using an array of glob patterns
	// forceCoverageMatch: [],

	// A path to a module which exports an async function that is triggered once before all test suites
	// globalSetup: undefined,

	// A path to a module which exports an async function that is triggered once after all test suites
	// globalTeardown: undefined,

	// A set of global variables that need to be available in all test environments
	// globals: {},

	// The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
	// maxWorkers: "50%",

	// An array of directory names to be searched recursively up from the requiring module's location
	// moduleDirectories: [
	//   "node_modules"
	// ],

	// An array of file extensions your modules use
	moduleFileExtensions: ["js", "ts", "json"],

	// A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
	// moduleNameMapper: {},

	// An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
	// modulePathIgnorePatterns: [],

	// Activates notifications for test results
	// notify: false,

	// An enum that specifies notification mode. Requires { notify: true }
	// notifyMode: "failure-change",

	// A preset that is used as a base for Jest's configuration
	// preset: undefined,

	// Run tests from one or more projects
	// projects: undefined,

	// Use this configuration option to add custom reporters to Jest
	// reporters: undefined,

	// Automatically reset mock state before every test
	// resetMocks: false,

	// Reset the module registry before running each individual test
	// resetModules: false,

	// A path to a custom resolver
	// resolver: undefined,

	// Automatically restore mock state and implementation before every test
	// restoreMocks: false,

	// The root directory that Jest should scan for tests and modules within
	// rootDir: "tests",

	// A list of paths to directories that Jest should use to search for files in
	roots: ["src"],

	// Allows you to use a custom runner instead of Jest's default test runner
	// runner: "jest-runner",

	// The paths to modules that run some code to configure or set up the testing environment before each test
	// setupFiles: [],

	// A list of paths to modules that run some code to configure or set up the testing framework before each test
	// setupFilesAfterEnv: [],

	// The number of seconds after which a test is considered as slow and reported as such in the results.
	// slowTestThreshold: 5,

	// A list of paths to snapshot serializer modules Jest should use for snapshot testing
	// snapshotSerializers: [],

	// The test environment that will be used for testing
	// testEnvironment: "jest-environment-node",

	// Options that will be passed to the testEnvironment
	// testEnvironmentOptions: {},

	// Adds a location field to test results
	// testLocationInResults: false,

	// The glob patterns Jest uses to detect test files
	testMatch: ["**/?(*.)+(spec|test).[tj]s?(x)"],

	// An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
	testPathIgnorePatterns: ["\\\\node_modules\\\\"],

	// The regexp pattern or array of patterns that Jest uses to detect test files
	// testRegex: [],

	// This option allows the use of a custom results processor
	// testResultsProcessor: undefined,

	// This option allows use of a custom test runner
	// testRunner: "jest-circus/runner",

	// A map from regular expressions to paths to transformers
	transform: {
		"^.+\\.(ts|tsx)$": "ts-jest",
	},

	// An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
	// transformIgnorePatterns: [
	//   "\\\\node_modules\\\\",
	//   "\\.pnp\\.[^\\\\]+$"
	// ],

	// An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
	// unmockedModulePathPatterns: undefined,

	// Indicates whether each individual test should be reported during the run
	// verbose: undefined,

	// An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
	// watchPathIgnorePatterns: [],

	// Whether to use watchman for file crawling
	// watchman: true,
}
